connector/keystone/types.go

```diff
@@ -0,0 +1,102 @@
+package keystone
+
+import (
+	"github.com/sirupsen/logrus"
+)
+
+type keystoneConnector struct {
+	Domain           string
+	KeystoneHost     string
+	KeystoneUsername string
+	KeystonePassword string
+	Logger           logrus.FieldLogger
+}
+
+type userKeystone struct {
+	Domain domainKeystone `json:"domain"`
+	ID     string         `json:"id"`
+	Name   string         `json:"name"`
+}
+
+type domainKeystone struct {
+	ID   string `json:"id"`
+	Name string `json:"name"`
+}
+
+// Config holds the configuration parameters for Keystone connector.
+// Keystone should expose API v3
+// An example config:
+//	connectors:
+//		type: keystone
+//		id: keystone
+//		name: Keystone
+//		config:
+//			keystoneHost: http://example:5000
+//			domain: default
+//      keystoneUsername: demo
+//      keystonePassword: DEMO_PASS
+type Config struct {
+	Domain           string `json:"domain"`
+	KeystoneHost     string `json:"keystoneHost"`
+	KeystoneUsername string `json:"keystoneUsername"`
+	KeystonePassword string `json:"keystonePassword"`
+}
+
+type loginRequestData struct {
+	auth `json:"auth"`
+}
+
+type auth struct {
+	Identity identity `json:"identity"`
+}
+
+type identity struct {
+	Methods  []string `json:"methods"`
+	Password password `json:"password"`
+}
+
+type password struct {
+	User user `json:"user"`
+}
+
+type user struct {
+	Name     string `json:"name"`
+	Domain   Domain `json:"domain"`
+	Password string `json:"password"`
+}
+
+// Domain struct holds an ID of domain
+type Domain struct {
+	ID string `json:"id"`
+}
+
+type token struct {
+	IssuedAt  string                 `json:"issued_at"`
+	Extras    map[string]interface{} `json:"extras"`
+	Methods   []string               `json:"methods"`
+	ExpiresAt string                 `json:"expires_at"`
+	User      userKeystone           `json:"user"`
+}
```

Comment:
I think we don't need to care about ExpiresAt - because we are using token immediately after it's obtained. Keystone tokens are quite short living (I think by default 1h). Keeping it in identity.ConnectorData and using for Refresh would limit liveness of dex refresh token by expiration of keystone user token. That's why we require keystone admin account credentials. Admin should have access to other's users and their groups. When we check refresh token, the admin account gets fresh token from keystone and verifies user existence and groups. I hope that it makes some sense :-)